File: src/main/java/dev/vrba/secrethitler/engine/EnactedPolicies.java
package dev.vrba.secrethitler.engine;

import dev.vrba.secrethitler.engine.election.PresidentialAbility;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

@Getter
@AllArgsConstructor
public class EnactedPolicies {

    private final int fascist;

    private final int liberal;

    private final Map<Integer, PresidentialAbility> abilities;

    public EnactedPolicies(int players) {
        this.fascist = 0;
        this.liberal = 0;
        this.abilities = PresidentialAbility.getAbilities(players);
    }

    public @NotNull EnactedPolicies add(@NotNull Party policy) {
        return switch (policy) {
            case LIBERAL -> new EnactedPolicies(fascist, liberal + 1, this.abilities);
            case FASCIST -> new EnactedPolicies(fascist + 1, liberal, this.abilities);

            default -> throw new IllegalArgumentException("Expected either liberal or fascist policy.");
        };
    }
}

File: src/main/kotlin/dev/vrba/secrethitler/engine/EnactedPolicies.kt
package dev.vrba.secrethitler.engine

const val requiredFascistPolicies = 6
const val requiredLiberalPolicies = 5

data class EnactedPolicies(val fascist: Int, val liberal: Int, val abilities: Map<Int, PresidentialAbility>) {

    constructor(players: Int) : this(0, 0, getPresidentialAbilities(players))

    fun addPolicy(policy: Policy): EnactedPolicies =
        when (policy) {
            Policy.Fascist -> EnactedPolicies(fascist + 1, liberal, abilities)
            Policy.Liberal -> EnactedPolicies(fascist, liberal + 1, abilities)
            else -> throw IllegalArgumentException("Expected either liberal or fascist policy.")
        }
}